<a href="https://colab.research.google.com/github/prajwalDU/customer-segmentation/blob/main/Online_Retail_Customer_Segmentation_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <b><u> Project Title : Extraction/identification of major topics & themes discussed in news articles. </u></b>

## <b> Problem Description </b>

### In this project, your task is to identify major customer segments on a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail.The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers.

## <b> Data Description </b>

### <b>Attribute Information: </b>

* ### InvoiceNo: Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.
* ### StockCode: Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.
* ### Description: Product (item) name. Nominal.
* ### Quantity: The quantities of each product (item) per transaction. Numeric.
* ### InvoiceDate: Invice Date and time. Numeric, the day and time when each transaction was generated.
* ### UnitPrice: Unit price. Numeric, Product price per unit in sterling.
* ### CustomerID: Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.
* ### Country: Country name. Nominal, the name of the country where each customer resides.

Importing python packages

In [25]:
# Importing library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [26]:
# Mounting drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [27]:
# Importing dataset
data = '/content/drive/MyDrive/almabetter/ML/customer segmentation/Online Retail.xlsx'

df = pd.read_excel(data)

In [28]:
# Check first 5 rows of dataset
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [39]:
# Check last 5 rows of dataset
df.tail()

InvoiceNo StockCode                      Description  Quantity  \
541904    581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID Country  
541904 2011-12-09 12:50:00       0.85     12680.0  France  
541905 2011-12-09 12:50:00       2.10     12680.0  France  
541906 2011-12-09 12:50:00       4.15     12680.0  France  
541907 2011-12-09 12:50:00       4.15     12680.0  France  
541908 2011-12-09 12:50:00       4.95     12680.0  France

In [29]:
# Check the shape of dataset
df.shape

(541909, 8)

In [30]:
# Check the describe of dataset
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

Check the duplicate values in our dataset

In [31]:
# Check the duplicates in dataset
len(df[df.duplicated()])

5268

We have 5268 duplicates values in our dataset we as to drop it before starting analysis

In [32]:
# Print the shape of dataset before removing duplicate
print('The shape of dataset before removing duplicate :', df.shape)

# Dropping duplicate
df.drop_duplicates(inplace=True)

# Print the shape of dataset after removing duplicate
print('The shape of dataset after removing duplicate :', df.shape)

The shape of dataset before removing duplicate : (541909, 8)
The shape of dataset after removing duplicate : (536641, 8)


Check the missing values in dataset with percentage

In [33]:
# Check the missing values
missing = df.columns[df.isnull().any()].tolist()

# Print the missing values in dataset
print('Missing Data count')
print(df[missing].isnull().sum().sort_index(ascending=False))
# Print the percentage of missing data
print('Missing Data Percentage')
print(round(df[missing].isnull().sum().sort_index(ascending=False)/len(df)*100,2))

Missing Data count
Description      1454
CustomerID     135037
dtype: int64
Missing Data Percentage
Description     0.27
CustomerID     25.16
dtype: float64


Dropping rows where customerID column contain null

In [37]:
# Dropping customerID where row with nulls
df.dropna(subset=['CustomerID'], inplace=True)

In [38]:
# Check shape of dataset
df.shape

(401604, 8)

In [35]:
# Check the info of dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 536641 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    536641 non-null  object        
 1   StockCode    536641 non-null  object        
 2   Description  535187 non-null  object        
 3   Quantity     536641 non-null  int64         
 4   InvoiceDate  536641 non-null  datetime64[ns]
 5   UnitPrice    536641 non-null  float64       
 6   CustomerID   401604 non-null  float64       
 7   Country      536641 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 36.8+ MB


**Feature Engineering**

AS we seen above we have one datetime column we take new columns from that like date, month, year, hour

In [41]:
df['year'] = df['InvoiceDate'].apply(lambda x : x.year)
df['month'] = df['InvoiceDate'].apply(lambda x : x.month_name())
df['day'] = df['InvoiceDate'].apply(lambda x : x.day_name())
df['hour'] = df['InvoiceDate'].apply(lambda x : x.hour)

creating another one feature total amount

In [42]:
df['TotalAmount'] = df['Quantity'] * df['UnitPrice']

Above we created hour columns for further analysis we divid time into 3 zone morning, afternoon and evening.

In [44]:
df['TimeType'] = np.where((df["hour"]>5)&(df["hour"]<18), np.where(
                           df["hour"]<12, 'Morning','Afternoon'),'Evening')

In [48]:
# Check cancellation using InvoiceNo
df['InvoiceNo'] = df['InvoiceNo'].astype('str')
cancellations = df[df['InvoiceNo'].str.contains('C')]
cancellations.head()

InvoiceNo StockCode                       Description  Quantity  \
141   C536379         D                          Discount        -1   
154   C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235   C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236   C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237   C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   

            InvoiceDate  UnitPrice  CustomerID         Country  year  \
141 2010-12-01 09:41:00      27.50     14527.0  United Kingdom  2010   
154 2010-12-01 09:49:00       4.65     15311.0  United Kingdom  2010   
235 2010-12-01 10:24:00       1.65     17548.0  United Kingdom  2010   
236 2010-12-01 10:24:00       0.29     17548.0  United Kingdom  2010   
237 2010-12-01 10:24:00       0.29     17548.0  United Kingdom  2010   

        month        day  hour  TotalAmount TimeType  
141  December  Wednesday     9       -27.50  Morning  
154  December  Wednesday     9        -4.65  Morning  
235  December  Wednesday    10       -19.80  Morning  
236  December  Wednesday    10        -6.96  Morning  
237  December  Wednesday    10        -6.96  Morning

In [49]:
# Check cancellations shape
cancellations.shape

(8872, 14)

We have total 401604 InvoiceNo in that we have 8872 cancellations data, so we has to drop it

In [50]:
# Drop the rows which contain 'C' in our main dataset
dt = df[~df['InvoiceNo'].str.contains('C')]

In [52]:
# Check our dataset shape
dt.shape

(392732, 14)

# **EDA**